File: go/Q5_Polimorfismo/Q5_Polimorfismo.go
package main

import (
	"fmt"
)

type Animal interface {
	Som() string
}

type AnimalBase struct {
	nome    string
	especie string
}

func (a AnimalBase) Som() string {
	return ""
}

type Cachorro struct {
	AnimalBase 
}

func (c Cachorro) Som() string {
	return "Au Au"
}

type Gato struct {
	AnimalBase
}

func (g Gato) Som() string {
	return "Miau"
}

func emitirSons(animais []Animal) {
	for _, a := range animais {
		fmt.Println(a.Som())
	}
}

func main() {
	cachorro := Cachorro{AnimalBase: AnimalBase{nome: "Bicho", especie: "Cachorro"}}
	gato := Gato{AnimalBase: AnimalBase{nome: "Bichano", especie: "Gato"}}

	fmt.Println(cachorro.Som())
	fmt.Println(gato.Som())

	animais := []Animal{cachorro, gato}
	emitirSons(animais)
}

File: Python/Q5_Polimorfismo.py
class Animal:
    def __init__(self,especie,nome):
        self.especie = especie
        self.nome = nome

    def emitir_som(self):
        pass

class Cachorro(Animal):
    def emitir_som(self):
        return "Au Au"

class Gato(Animal):
    def emitir_som(self):
        return "Miau"

def emitir_sons(*animais):
    for i in range(len(animais)):
        print(animais[i].emitir_som())
        
bicho = Cachorro(Cachorro, "Bicho")
bicho2 = Gato(Gato, "Bichano")
print(bicho2.emitir_som())
emitir_sons(bicho,bicho2)